### DISCUSSION TITLE
Periodic boundary condition for an RVE sub-model

### DISCUSSION THREAD

**Discussion by abarun22** [idaholab/moose#18380] [Q&A General] [answered]
My current work focus on modelling a material microstructure on an RVE and eventually embedding the homogenized behaviour of this model on to bigger component models in continuum scale. As a first step i am interested in applying peridoc boundary conditions to this RVE and i was exploring various methods available in MOOSE which can do that job. The AddPeriodicBCAction objects contain straight forward methods to apply this BC, however to my dismay, the results were not on the expected lines. All of the related methods (auto, translation, function transformation) were tried. 

Alternatively i tried an another route in the form of constraints which might as well model the periodic boundary conditions. This however does not allow me to reference the two pairs of edges participating in this model. Seems like the meshgenerator would either want one pair of edges to be defined or definition of "final_generator" parameter which i could not understand, especially for this single domain model. 

A look in to the test suite (/test/tests/mortar/periodic-value/gold) suggests that this approach might give closer results, provided we reference the boundaries to be matched correctly. Here is what i received.

```
*** ERROR ***
Your MeshGenerator tree contains multiple possible generator outputs :
"master_y and one or more of the following from an independent set: "master_x, slave_y"
```

This may be due to a missing dependency or may be intentional. Please select the final MeshGenerator in
the [Mesh] block with the "final_generator" parameter or add additional dependencies to remove the ambiguity.

As far my submodel is concerned it is a simple 2D square block with normal displacements constrained in two of the edges and displacement loads applied on the right edge. The application of PBC's must ensure that the displacements on the respective opposite edges are the same and that we get the simlar deformed profiles on these edges. At present i could not get the sort of behaviour one would expect realistically for this kind of model. e.g., similar to the results in the above folder. I really appreciate any suggestion or pointers in this regard, that would provide a solid start to the work. The attached files might give you an insight in to the schematic of the model, input structure and the results obtained so far.
[19072021.zip](https://github.com/idaholab/moose/files/6842971/19072021.zip)

Kind regards,
Arun

**Comment by abarun22** [comment]
A further update to my earlier query on cracking the PBC application. Here is an improved version of the input file for mortar based PBC, where the input meshes (for the object 'LowerDBlockFromSidesetGenerator') were given in sequence for each extraction of block of elements in the edge pairs. This has cured the problem of multiple mesh generator outputs that i mentioned before. Following is the modified version of input commands used.

```
[Mesh]
  [file]
    type = FileMeshGenerator
    file = Square.e
  []
  [slave_x]
    input = file
    type = LowerDBlockFromSidesetGenerator
    new_block_id = 11
    new_block_name = "secondary_x"
    sidesets = 'right_edge'
  []
  [master_x]
    input = slave_x
    type = LowerDBlockFromSidesetGenerator
    new_block_id = 12
    new_block_name = "primary_x"
    sidesets = 'left_edge'
  []
  [slave_y]
    input = master_x
    type = LowerDBlockFromSidesetGenerator
    new_block_id = 13
    new_block_name = "secondary_y"
    sidesets = 'bottom_edge'
  []
  [master_y]
    input = slave_y
    type = LowerDBlockFromSidesetGenerator
    new_block_id = 14
    new_block_name = "primary_y"
    sidesets = 'top_edge'
  []
[]

[GlobalParams]
  displacements = 'disp_x disp_y'
[]

[Variables]
  [lmx]
    block = "secondary_x"
  []
  [lmy]
    block = "secondary_y"
  []
[]

[Variables]
  [./disp_x]
    scaling = 1e-10
  [../]
  [./disp_y]
    scaling = 1e-10
  [../]
[]

[Kernels]
  [./stress_x]
    type = ADStressDivergenceTensors
    component = 0
    variable = disp_x
    use_displaced_mesh = true
  [../]
  [./stress_y]
    type = ADStressDivergenceTensors
    component = 1
    variable = disp_y
    use_displaced_mesh = true
  [../]
[]

[BCs]
  [./symmty]
      type = DirichletBC
      variable = disp_y
      boundary = "top_edge"
      value = 0.0
  [../]

  [./symmlx]
      type = DirichletBC
     variable = disp_x
      boundary = "left_edge"
      value = 0.0
  [../]

  [./dispR]
    type = FunctionDirichletBC
    variable = disp_x
    function = '0.1*t'
    boundary = "right_edge"
  [../]
[]

[Constraints]
  [cx]
    type = EqualValueConstraint
    variable = lmx
    secondary_variable = disp_x
    primary_boundary = "left_edge"
    secondary_boundary = "right_edge"
    primary_subdomain = 12
    secondary_subdomain = 11
    periodic = true
  []
  [cy]
    type = EqualValueConstraint
    variable = lmy
    secondary_variable = disp_y
    primary_boundary = "top_edge"
    secondary_boundary = "bottom_edge"
    primary_subdomain = 14
    secondary_subdomain = 13
    periodic = true
  []
[]


[AuxVariables]
  [./stress_xx]
    order = CONSTANT
    family = MONOMIAL
  [../]
  [./stress_yy]
    order = CONSTANT
    family = MONOMIAL
  [../]
  [./von_mises]
    #Dependent variable used to visualize the Von Mises stress
    order = CONSTANT
    family = MONOMIAL
  [../]
[]

[AuxKernels]
  [./str_xx]
    type = ADRankTwoAux
    rank_two_tensor = stress
    variable = stress_xx
    index_i = 0
    index_j = 0
  [../]

  [./str_yy]
    type = ADRankTwoAux
    rank_two_tensor = stress
    variable = stress_yy
    index_i = 1
    index_j = 1
  [../] 
  
  [./von_mises_kernel]
    #Calculates the von mises stress and assigns it to von_mises
    type = ADRankTwoScalarAux
    variable = von_mises
    rank_two_tensor = stress
    execute_on = timestep_end
    scalar_type = VonMisesStress
  [../]
[]
[Materials]
  [./elasticity]
    type = ADComputeIsotropicElasticityTensor
    poissons_ratio = 0.3
    youngs_modulus = 1e5
  [../]
  [./strain]
    type = ADComputeFiniteStrain
  [../]
  [./stress]
    type = ADComputeFiniteStrainElasticStress
  [../]
[]

[Postprocessors]
  [./stress_xx]
    type = ElementAverageValue
    variable = stress_xx
  [../]
  [./stress_yy]
    type = ElementAverageValue
    variable = stress_yy
 [../]
[]

[Preconditioning]
  [./smp]
    type = SMP
    full = true
  [../]
[]

[Executioner]
  num_steps = 50
  dt = 0.1
  type = Transient

 nl_rel_tol = 1e-6
 nl_max_its = 10
  l_tol = 1e-4
  l_max_its = 50

  solve_type = Newton
  petsc_options_iname = '-pc_type'
  petsc_options_value = 'lu'
[]

[Outputs]
  exodus = true
  csv = true
[]
```

This time i am hit with zero pivoting issues during the solution stage, that could possibly be related to rigid body displacements happening in the model. 

```
_**Time Step 1, time = 0.05, dt = 0.05
 0 Nonlinear |R| = 1.541104e-03
      0 Linear |R| = 1.541104e-03
  Linear solve did not converge due to DIVERGED_PC_FAILED iterations 0
                 PC failed due to FACTOR_NUMERIC_ZEROPIVOT
Nonlinear solve did not converge due to DIVERGED_LINE_SEARCH iterations 0
 Solve Did NOT Converge!
Aborting as solve did not converge

Time Step 1, time = 0.025, dt = 0.025
 0 Nonlinear |R| = 7.705519e-04
      0 Linear |R| = 7.705519e-04
  Linear solve did not converge due to DIVERGED_PC_FAILED iterations 0
                 PC failed due to FACTOR_NUMERIC_ZEROPIVOT
Nonlinear solve did not converge due to DIVERGED_LINE_SEARCH iterations 0
 Solve Did NOT Converge!
Aborting as solve did not converge**_
```

The model treats all the boundaries to be restrained in their respective normal directions and the application of constraints does not seem to resolve this, while the application of displacement controlled loads does not seem to deform the model. Kindly request you please to pour in your comments that would enable to me check if there are any deficiencies in the current input structure.
Kind regards,
Arun

**Comment by dschwen** [comment, marked as answer]
Your are overcomplicating PBCs. The `[Periodic]` action should be sufficient here. However it requires nodes to line up across periodic boundaries. Make sure your mesh is constructed accordingly (`GeneratedMeshGenerator` does it right by default).

**Reply by abarun22** [reply]
Thanks Daniel. Yes you are right. It works now, however with modified BC's as given below. This gives enough room for the model to deform in the loading direction preventing the RBM encountered earlier. 
![image](https://user-images.githubusercontent.com/40111479/126687433-8abca2db-3cfe-4ff3-bb6b-6cb5ca8770b3.png)

![image](https://user-images.githubusercontent.com/40111479/126687493-e6552abf-85c5-498e-ae4b-0fb906f6a218.png)

I hope the BC applied in this manner mimics realistically the response obtained for an infinite body, while integrating the results with larger continuum models.
Kind regards,
Arun

**Reply by hugary1995** [reply]
@abarun22 Could you paste your PBC block here? From the figure, it seems that you have some incompatible BCs.

**Reply by abarun22** [reply]
Dear Gary,
Here is the block for boundary conditions and also attached the new model file.
[BCs]
  [./Periodic]
    [./all]
      variable = 'disp_x disp_y'
      auto_direction = 'x y'
    [../]
  [../]

  [./top_x]
      type = DirichletBC
      variable = disp_x
      boundary = "left_top"
      value = 0.0
  [../]

  [./lbot_x]
      type = DirichletBC
      variable = disp_x
      boundary = "left_bot"
      value = 0.0
  [../]

  [./lbot_y]
      type = DirichletBC
      variable = disp_y
      boundary = "left_bot"
      value = 0.0
  [../]

  [./rbot_x]
      type = DirichletBC
      variable = disp_x
      boundary = "right_bot"
      value = 0.0
  [../]

  [./dispLoad]
    type = FunctionDirichletBC
    variable = disp_x
    function = '0.1*t'
    boundary = "right_edge"
  [../]
[]
[Square.zip](https://github.com/idaholab/moose/files/6867913/Square.zip)

**Reply by hugary1995** [reply]
For example, the rbot_x and dispLoad are incompatible at the bottom right corner.

**Reply by abarun22** [reply]
Yes i see that there is a problem of resolution at the bottom right corner. I guess the preference is given to the effect of restraint rather than the application of loads and therefore you see no displacements there. I guess with a denser mesh the kink at the top and bottom displacement profiles would be not that obvious. I am not pretty sure whether we are on the same lines.

**Reply by abarun22** [reply]
Further problems were observed when the model is run with a refined mesh. The periodicity effects are subdued and i see a variation in displacement profile along the boundaries. Attached please find the displacement contour along the same loading direction, that shows this effect. I guess there is certainly a problem here, as i do not see a dependency with respect to the FE model size. Any suggestions on fixing this issue would be highly appreciated.

![image](https://user-images.githubusercontent.com/40111479/126898832-1d683be9-a37d-4fb4-84b1-753c9cc1a482.png)

**Reply by abarun22** [reply]
Any suggestions/ideas on fixing this issue with refined mesh?

**Reply by dschwen** [reply]
You need to look at the Global Strain system https://mooseframework.inl.gov/source/actions/GlobalStrainAction.html will will ensure _periodic strains_! That's the right approach fro an RVE. Your loading will have to be formulated in terms of an applied stress to the RVE.

**Reply by abarun22** [reply]
Thanks Daniel. Thanks for pointing me to the right direction. Indeed this would fit my requirements very much and i hope to extend my developments further. I wanted to ask whether the framework would with external mesh combined with automatic mode of PBC application. Executing this i received the following error.
*** ERROR ***
Couldn't auto-detect a paired boundary for use with periodic boundary conditions

Looks like there is a issue identifying the periodic boundaries. Here is the input file, i've used for your reference.

[Mesh]
  [file]
    type = FileMeshGenerator
    file = cube.inp
  []
  [cnode]
    type = ExtraNodesetGenerator
    coord = '0.0 0.0 0.0'
    new_boundary = 100
    input = file
  []
[]

[Variables]
  [./u_x]
  [../]
  [./u_y]
  [../]
  [./u_z]
  [../]
  [./global_strain]
    order = SIXTH
    family = SCALAR
  [../]
[]

[AuxVariables]
  [./disp_x]
  [../]
  [./disp_y]
  [../]
  [./disp_z]
  [../]
  [./s00]
    order = CONSTANT
    family = MONOMIAL
  [../]
  [./s11]
    order = CONSTANT
    family = MONOMIAL
  [../]
  [./e00]
    order = CONSTANT
    family = MONOMIAL
  [../]
  [./e11]
    order = CONSTANT
    family = MONOMIAL
  [../]
[]

[AuxKernels]
  [./disp_x]
    type = GlobalDisplacementAux
    variable = disp_x
    scalar_global_strain = global_strain
    global_strain_uo = global_strain_uo
    component = 0
  [../]
  [./disp_y]
    type = GlobalDisplacementAux
    variable = disp_y
    scalar_global_strain = global_strain
    global_strain_uo = global_strain_uo
    component = 1
  [../]
  [./disp_z]
    type = GlobalDisplacementAux
    variable = disp_z
    scalar_global_strain = global_strain
    global_strain_uo = global_strain_uo
    component = 2
  [../]
  [./s00]
    type = RankTwoAux
    variable = s00
    rank_two_tensor = stress
    index_i = 0
    index_j = 0
  [../]
  [./s11]
    type = RankTwoAux
    variable = s11
    rank_two_tensor = stress
    index_i = 1
    index_j = 1
  [../]
  [./e00]
    type = RankTwoAux
    variable = e00
    rank_two_tensor = total_strain
    index_i = 0
    index_j = 0
  [../]
  [./e11]
    type = RankTwoAux
    variable = e11
    rank_two_tensor = total_strain
    index_i = 1
    index_j = 1
 [../]
[]

[GlobalParams]
  displacements = 'u_x u_y u_z'
[]

[Kernels]
  [./TensorMechanics]
  [../]
[]

[ScalarKernels]
  [./global_strain]
    type = GlobalStrain
    variable = global_strain
    global_strain_uo = global_strain_uo
  [../]
[]

[BCs]
  [./Periodic]
    [./au]
      variable = ' u_x u_y u_z'
      auto_direction = 'x y z'
    [../]
#    [./LR]
#      variable = 'u_x u_y u_z'
#      primary = "left"
#      secondary = "right"
#      translation = '1 0 0'
#    [../]
#    [./TB]
#      variable = 'u_x u_y u_z'
#      primary = "top"
#      secondary = "bottom"
#      translation = '0 1 0'
#    [../]
#    [./FB]
#      variable = 'u_x u_y u_z'
#      primary = "front"
#      secondary = "back"
#      translation = '0 0 1'
#    [../]
  [../]

  # fix center point location
  [./centerfix_x]
    type = DirichletBC
    boundary = 100
    variable = u_x
    value = 0
  [../]
  [./centerfix_y]
    type = DirichletBC
    boundary = 100
    variable = u_y
    value = 0
  [../]
  [./centerfix_z]
    type = DirichletBC
    boundary = 100
    variable = u_z
    value = 0
  [../]
[]

[Materials]
  [./elasticity_tensor]
    type = ComputeElasticityTensor
    C_ijkl = '70e9 0.33'
    fill_method = symmetric_isotropic_E_nu
  [../]
  [./strain]
    type = ComputeSmallStrain
    global_strain = global_strain
  [../]
  [./global_strain]
    type = ComputeGlobalStrain
    scalar_global_strain = global_strain
    global_strain_uo = global_strain_uo
  [../]
  [./stress]
    type = ComputeLinearElasticStress
  [../]
[]

[UserObjects]
  [./global_strain_uo]
    type = GlobalStrainUserObject
    applied_stress_tensor = '5e9 0 0 0 0 0'
    execute_on = 'Initial Linear Nonlinear'
  [../]
[]

[Postprocessors]
  [./l2err_e00]
    type = ElementL2Error
    variable = e00
    function = 0.07142857 #strain_xx = C1111/sigma_xx
  [../]
  [./l2err_e11]
    type = ElementL2Error
    variable = e11
    function = -0.07142857*0.33 #strain_yy = -nu*strain_xx
  [../]
[]

[Preconditioning]
  [./SMP]
    type = SMP
    full = true
  [../]
[]

[Executioner]
  type = Transient
  scheme = bdf2
  solve_type = 'PJFNK'

  line_search = basic

  petsc_options_iname = '-pc_type -ksp_gmres_restart -sub_ksp_type -sub_pc_type -pc_asm_overlap'
  petsc_options_value = 'asm         31   preonly   lu      1'

  l_max_its = 30
  nl_max_its = 12

  nl_rel_tol = 1.0e-10

  start_time = 0.0
  num_steps = 5
[]

[Outputs]
  exodus = true
[]
Instead i tried to use the translational PBC, (ones which are under commented section in BCs) which did not produce any results (pls refer the results pased below). 
Framework Information:
MOOSE Version:           git commit b06f1d9 on 2016-03-01
LibMesh Version:         c70ab40900490f44391f1b891c8c3731c45f0cf1
PETSc Version:           3.14.2
SLEPc Version:           3.14.0
Current Time:            Mon Aug  2 19:14:28 2021
Executable Timestamp:    Wed Jul 28 17:30:08 2021

Parallelism:
  Num Processors:          1
  Num Threads:             1

Mesh:
  Parallel Type:           replicated
  Mesh Dimension:          3
  Spatial Dimension:       3
  Nodes:
    Total:                 1331
    Local:                 1331
  Elems:
    Total:                 1000
    Local:                 1000
  Num Subdomains:          4
  Num Partitions:          1

Nonlinear System:
  Num DOFs:                3999
  Num Local DOFs:          3999
  Variables:               { "u_x" "u_y" "u_z" } "global_strain"
  Finite Element Types:    "LAGRANGE" "SCALAR"
  Approximation Orders:    "FIRST" "SIXTH"

Auxiliary System:
  Num DOFs:                7993
  Num Local DOFs:          7993
  Variables:               { "disp_x" "disp_y" "disp_z" } { "s00" "s11" "e00" "e11" }
  Finite Element Types:    "LAGRANGE" "MONOMIAL"
  Approximation Orders:    "FIRST" "CONSTANT"

Execution Information:
  Executioner:             Transient
  TimeStepper:             ConstantDT
  Solver Mode:             Preconditioned JFNK
  MOOSE Preconditioner:    SMP

LEGACY MODES ENABLED:
 This application uses the legacy material output option: material properties are output only on TIMESTEP_END, not INITIAL. To                                                                remove this message, set 'use_legacy_material_output' to false in this application. If there are gold output files that conta                                                               in material property output for which output occurs on INITIAL, then these will generate diffs due to zero values being stored                                                               , and these tests should be re-golded.
Time Step 0, time = 0

Postprocessor Values:
+----------------+----------------+----------------+
| time           | l2err_e00      | l2err_e11      |
+----------------+----------------+----------------+
|   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |
+----------------+----------------+----------------+


Scalar Variable Values:
+----------------+----------------+----------------+----------------+----------------+----------------+----------------+
| time           | global_strain_0| global_strain_1| global_strain_2| global_strain_3| global_strain_4| global_strain_5|
+----------------+----------------+----------------+----------------+----------------+----------------+----------------+
|   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |
+----------------+----------------+----------------+----------------+----------------+----------------+----------------+


Time Step 1, time = 1, dt = 1
 0 Nonlinear |R| = 0.000000e+00
Nonlinear solve converged due to CONVERGED_FNORM_ABS iterations 0
 Solve Converged!

Postprocessor Values:
+----------------+----------------+----------------+
| time           | l2err_e00      | l2err_e11      |
+----------------+----------------+----------------+
|   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |
|   1.000000e+00 |   7.142857e-02 |   2.357143e-02 |
+----------------+----------------+----------------+


Scalar Variable Values:
+----------------+----------------+----------------+----------------+----------------+----------------+----------------+
| time           | global_strain_0| global_strain_1| global_strain_2| global_strain_3| global_strain_4| global_strain_5|
+----------------+----------------+----------------+----------------+----------------+----------------+----------------+
|   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |
|   1.000000e+00 |   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |
+----------------+----------------+----------------+----------------+----------------+----------------+----------------+
I guess this formulation based load application will not be done discretely and so the reason for a single time step given in the executioner block. Please could you advise me whether "GeneratedMesh" is the only option to execute problems with PBC's applied in an automatic fashion? And offcourse, any suggestions to make the translational method to work would be much welcomed.

Attached here is the ABAQUS mesh used for this model, if you would like to give a check.
[cube.txt](https://github.com/idaholab/moose/files/6922989/cube.txt)

Kind regards,
Arun
